feat(vllm): Add prompt embeds support for pre-computed inference inputs

wenqiglantz opened the pull request:
#### Overview:

This PR adds prompt embeds (also known as prompt embeddings) as a secure alternative input method to traditional text prompts in Dynamo. By allowing applications to use pre-computed embeddings for inference, this feature not only offers greater flexibility in prompt engineering but also significantly enhances privacy and data security. With prompt embeddings, sensitive user data can be transformed into embeddings before ever reaching the inference server, reducing the risk of exposing confidential information during the AI workflow.

#### Details:

Key changes:

- API: Added `prompt_embeds` field to completion requests (base64-encoded PyTorch tensor)
- Frontend (Rust): Validates base64 encoding, size limits (100B-10MB), skips tokenization when embeddings present
- Worker (Python): Decodes PyTorch tensors, creates EmbedsPrompt for vLLM, extracts sequence length for usage stats
- Usage Stats: Fixed prompt_tokens reporting—worker extracts from tensor shape, frontend propagates correctly
- NATS: Increased max_payload to 15MB to accommodate 10MB decoded embeddings
- Docs: Added `docs/backends/vllm/prompt-embeddings.md`

#### Where should the reviewer start?

- `components/src/dynamo/vllm/handlers.py` — Core worker implementation:
  - `_decode_prompt_embeds()` — PyTorch tensor decoding
  - `DecodeWorkerHandler.generate()` — Embeddings path (lines ~372-420)
  - `_build_completion_usage()` — Usage stats with `embedding_sequence_length`
- `lib/llm/src/protocols/openai/validate.rs` — Validation functions:
  - `validate_prompt_embeds()` — Base64 + size validation
  - `validate_prompt_or_embeds()` — At-least-one check
- `lib/llm/src/preprocessor.rs` — Frontend handling
- `lib/llm/src/protocols/openai/completions/delta.rs` — Usage propagation fix
- `components/src/dynamo/vllm/tests/test_prompt_embeds.py` — Unit tests for decoder and usage stats

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- closes GitHub issue: #xxx


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added prompt embeddings support to the vLLM backend as an alternative to text prompts
  * New `--enable-prompt-embeds` CLI flag to enable the feature
  * New `prompt_embeds` field in completion requests accepting base64-encoded tensors (max 10MB decoded)
  * Embeddings take precedence when both prompts and embeddings are provided

* **Configuration**
  * Increased NATS JetStream maximum payload size from 10MB to 15MB

* **Documentation**
  * Added comprehensive guides for prompt embeddings usage, API reference, and deployment examples

* **Tests**
  * Added extensive unit and integration tests for prompt embeddings functionality

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

wenqiglantz committed:
feat(vllm): Add prompt embeds support for pre-computed inference inputs

- Add prompt_embeds field to completion API for base64-encoded PyTorch tensors
- Decode and validate embeddings in vLLM worker handlers
- Skip tokenization when embeddings provided (embeddings take precedence)
- Update NATS payload limit to 15MB for large embeddings
- Add comprehensive documentation and integration tests
- Support accurate usage statistics for embedding-based requests

Signed-off-by: Wenqi Glantz <wglantz@nvidia.com>

wenqiglantz committed:
fix: fix mypy error

Signed-off-by: Wenqi Glantz <wglantz@nvidia.com>

grahamking commented:
/ok to test 19b9aa8

grahamking commented on lib/llm/src/preprocessor.rs:
```diff
@@ -1030,13 +1036,35 @@ impl
+        if request.inner.prompt_embeds.is_none() {
+            let isl = common_request.token_ids.len() as u32;
+            response_generator.update_isl(isl);
+        }
```
Is there a way to put both `update_isl` calls nearer to each other in the code? The other is line 900-ish above. That would make it easier to read and change.

wenqiglantz replied:
Good point. Extracted the ISL update logic into a helper method, defined once and called from both places.

wenqiglantz commented on components/src/dynamo/vllm/handlers.py:
```diff
@@ -777,10 +897,54 @@ async def generate(self, request, context):
+            token_ids = request["token_ids"]
+            prompt = TokensPrompt(
+                prompt_token_ids=token_ids, multi_modal_data=multi_modal_data
+            )
```
Good catch! Fixed it.

ayushag-nv replied:
just thinking if it will make sense to just wrap entire if else under a logical method ?? I am fine with the current state as well but that is duplication and devs have to be careful to change at both places.

wenqiglantz commented on lib/async-openai/src/types/completion.rs:
```diff
@@ -104,6 +104,13 @@ pub struct CreateCompletionRequest {
+    /// If both are provided, prompt_embeds takes precedence.
+    /// Maximum size: 10MB decoded.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub prompt_embeds: Option<String>,
```
It's a String. The `prompt_embeds` field contains base64-encoded binary data, not raw floats.  The data flow is:

```
Client                           API                              Backend
───────                         ─────                            ─────────
torch.Tensor (floats)  →  base64 encode  →  JSON String  →  base64 decode  →  torch.Tensor
[0.1, 0.2, ...]              "UEsDBBQ..."     "UEsDBBQ..."        bytes           [0.1, 0.2, ...]
```

wenqiglantz commented on lib/llm/src/protocols/openai/completions.rs:
```diff
@@ -494,4 +511,178 @@ mod tests {
+    }
+
+    #[test]
+    fn test_embeds_too_small() {
```
The 100-byte minimum is a sanity check to catch obviously invalid data. Real embeddings are much larger. 

PyTorch overhead:
- Magic bytes + version: ~10 bytes
- Metadata (shape, dtype, strides): ~50-100 bytes
- Minimum realistic file: hundreds of bytes

What 100 bytes catches:
- Empty/accidental data
- Truncated/corrupted uploads
- Wrong data type accidentally encoded
- User mistakes (sending text instead of tensor)

wenqiglantz commented on lib/llm/src/preprocessor.rs:
```diff
@@ -1030,13 +1036,35 @@ impl
+            builder.prompt_embeds(Some(prompt_embeds.clone()));
+
+            // Still gather multimodal data (embeddings can have images/video)
+            self.gather_multi_modal_data(&request, &mut builder).await?;
```
Good catch.  Fixed it.

wenqiglantz commented on tests/integration/__init__.py:
```diff
@@ -0,0 +1,2 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
```
This is where I needed your guidance. :-)  Fixed per your recommendation.

wenqiglantz commented on tests/integration/test_prompt_embeds_integration.py:
```diff
@@ -0,0 +1,627 @@
+@pytest.mark.nightly  # Run in nightly CI
+@pytest.mark.gpu_1  # Requires 1 GPU
+@pytest.mark.model(TEST_MODEL)
+class TestPromptEmbedsUsageStatistics:
```
see above comment

wenqiglantz commented on components/src/dynamo/vllm/handlers.py:
```diff
@@ -675,9 +751,52 @@ async def generate(self, request, context):
+                # Use EmbedsPrompt to create the correct input dict for vLLM 0.11.0
+                # EmbedsInputs TypedDict has: {type: 'embeds', prompt_embeds: Tensor, cache_salt?: str}
+                # It does NOT support prompt_token_ids
+                from vllm.inputs import EmbedsPrompt
```
Moved it to the top

wenqiglantz committed:
fix: Address PR review feedback for prompt_embeds

Signed-off-by: Wenqi Glantz <wglantz@nvidia.com>

wenqiglantz committed:
fix: format issues and CI error

Signed-off-by: Wenqi Glantz <wglantz@nvidia.com>

ayushag-nv commented on lib/llm/src/protocols/openai/completions.rs:
```diff
@@ -416,7 +416,22 @@ impl ValidateRequest for NvCreateCompletionRequest {
+        }
+
+        // Validate prompt_embeds if provided
+        validate::validate_prompt_embeds(self.inner.prompt_embeds.as_deref())?;
```
instead of 3 conditions. Can we just do 1 and check what is given and what is not ?? 

can we just use `validate_prompt_or_embeds`

wenqiglantz replied:
Fixed. Consolidated prompt validation into single function.

ayushag-nv commented on deploy/docker-compose.yml:
```diff
@@ -11,7 +11,9 @@ networks:
-    command: [ "-js", "--trace", "-m", "8222" ]
+    # max_payload set to 15MB to accommodate 10MB decoded embeddings
+    # Base64 encoding: 10MB → ~13.3MB, 15MB provides ~1.7MB buffer for metadata
+    command: [ "-js", "--trace", "-m", "8222", "--max_payload", "15728640" ]
```
Is the system broken without this change ? has this been tested ??

ayushag-nv commented on lib/llm/src/protocols/openai/chat_completions/delta.rs:
```diff
@@ -319,14 +319,17 @@ impl crate::protocols::openai::DeltaGeneratorExt<NvCreateChatCompletionStreamRes
+            // the embedding sequence length computed by the worker
+            if let Some(completion_usage) = delta.completion_usage.as_ref() {
+                // Update prompt_tokens from worker if provided (e.g., for embeddings)
+                self.usage.prompt_tokens = completion_usage.prompt_tokens;
```
I am wondering what do these two fields represent ? 
- prompt_tokens
- promt_tokens_details

ayushag-nv commented on components/src/dynamo/vllm/tests/test_vllm_prompt_embeds.py:
There are so many tests added for this here ? possibility to condense into fewer number of tests ?

wenqiglantz committed:
Merge main into prompt_embeds branch

wenqiglantz committed:
refactor: Extract LoRA-aware logging into helper method

Address code review feedback by consolidating duplicated if-else
logging patterns into a reusable _log_with_lora_context() method.
This reduces code duplication and ensures consistent logging format
across generate_tokens() and prefill handlers.

Signed-off-by: Wenqi Glantz <wglantz@nvidia.com>

wenqiglantz committed:
refactor: Move imports to module level in handlers.py

Address code review feedback by moving base64, binascii, io, and torch
imports from inside _decode_prompt_embeds() to the top of the file.
This follows Python best practices of having imports at module level.

Signed-off-by: Wenqi Glantz <wglantz@nvidia.com>

wenqiglantz committed:
refactor: Extract prompt building logic into shared method

Address code review feedback by consolidating duplicated prompt_embeds
vs token_ids handling logic from DecodeWorkerHandler._generate_token_mode
and PrefillWorkerHandler._generate_token_mode into a shared
BaseWorkerHandler._build_prompt_from_request() method.

This eliminates code duplication and ensures developers only need to
maintain one place for the prompt building logic.

Signed-off-by: Wenqi Glantz <wglantz@nvidia.com>

wenqiglantz committed:
refactor: Consolidate prompt validation into single function

Address code review feedback by combining 3 separate validation calls
into a single validate_prompt_or_embeds() function that:
- Checks at least one of prompt or prompt_embeds is provided
- Validates prompt_embeds format if provided (base64, size limits)
- Validates prompt content only if prompt_embeds is NOT provided

This simplifies the call site from 3 conditions to 1.

Signed-off-by: Wenqi Glantz <wglantz@nvidia.com>

wenqiglantz committed:
refactor: Condense prompt_embeds tests using pytest.parametrize

Address code review feedback by combining similar tests:
- Shape tests (2D/3D/dtypes) â†’ 1 parametrized test
- Invalid input tests â†’ 1 parametrized test
- Value range tests â†’ 1 parametrized test
- Usage stats tests â†’ 1 parametrized test + 2 edge cases

Reduced from 17 tests to 11 tests while maintaining full coverage.
Added shared fixtures and helper functions for cleaner code.

Signed-off-by: Wenqi Glantz <wglantz@nvidia.com>

wenqiglantz committed:
fix: format issue

Signed-off-by: Wenqi Glantz <wglantz@nvidia.com>

wenqiglantz committed:
Merge main into prompt_embeds branch

Signed-off-by: Wenqi Glantz <wglantz@nvidia.com>

wenqiglantz committed:
fix: update copyright year to 2026 in test files

Signed-off-by: Wenqi Glantz <wglantz@nvidia.com>

ayushag-nv reviewed (approved):
Looks good to me. @grahamking can additionally approve and resolve his comments

wenqiglantz committed:
Merge latest main into prompt_embeds

Signed-off-by: Wenqi Glantz <wglantz@nvidia.com>

wenqiglantz committed:
Merge remote-tracking branch 'origin/main' into prompt_embeds

wenqiglantz committed:
Merge remote-tracking branch 'origin/main' into prompt_embeds

biswapanda reviewed (approved):
lgtm
